File: Mentorierte_Arbeit_Induktion_Rekursion/n_queen_problem.py
import numpy as np

def check_square(row, column, board, n):
    # check if column is already occupied by a queen in one of the rows above
    i = row - 1
    while (i >= 0):
        if board[i][column] == 1:
            return False
        i -= 1

    # check north-western diagonal
    i = row - 1
    j = column - 1
    while ((i >= 0) and (j >= 0)):
        if board[i][j] == 1:
            return False
        i -= 1
        j -= 1

    # check north-eastern diagonal
    i = row - 1
    j = column + 1
    while ((i >= 0) and (j < n)):
        if board[i][j] == 1:
            return False
        i -= 1
        j += 1
    
    return True

number_of_solutions = 0
def solve_queen_problem(board, n, row = 0):
    global number_of_solutions # can be done way prettier! for example with classes
    if (row == n):
        number_of_solutions += 1
        print(board)
        return
    
    for j in range(n):
        if (check_square(row, j, board, n)):
            board[row][j] = 1
            solve_queen_problem(board, n, row + 1)
            board[row][j] = 0

def run():
    n = 8
    board = np.zeros((n,n), dtype = int)
    solve_queen_problem(board, n)
    print("for n = {0:>2}".format(n), ", there are exactly", "{0:>15}".format(number_of_solutions), " solutions", sep = '')

run()

File: Mentorierte_Arbeit_Induktion_Rekursion/src/n_queen_problem.cpp
// n_queen_problem.cpp

#include <iostream>
#include <cstring>
#include "n_queen_problem.h"

Queen_Problem::Queen_Problem(const unsigned int board_size) : number_of_solutions{0}, n{board_size}
{
	board = new bool*[n];
	for (unsigned int i = 0; i < n; ++i)
	{
		board[i] = new bool[n];
		memset(board[i], false, n * sizeof(bool));
	}
}

void Queen_Problem::print_solution(void) const
{
	std::cout << "\n";
	for (unsigned int i = 0; i < n; ++i)
	{
		for (unsigned int j = 0; j < n; ++j)
		{
			std::cout << board[i][j] << " ";
		}
		std::cout << "\n";
	}
	std::cout << "\n";
}

bool Queen_Problem::check_square(const unsigned int row, const unsigned int column) const
{
	int i;
	int j;

	// check if column is already occupied by a queen in one of the rows above
	for (i = row - 1; i >= 0; --i)
	{
		if (board[i][column] == 1)
		{
			return false;
		}
	}

	// check north-western diagonal
	for (i = row - 1, j = column - 1; (i >= 0) and (j >= 0); --i, --j)
	{
		if (board[i][j] == 1)
		{
			return false;
		}
	}

	// check north-eastern diagonal
	for (i = row - 1, j = column + 1; (i >= 0) and (j < static_cast<int>(n)); --i, ++j)
	{
		if (board[i][j] == 1)
		{
			return false;
		}
	}
	return true;
}

void Queen_Problem::solve_queen_problem(const unsigned int row)
{
	if (row == n)
	{
		print_solution();
		++number_of_solutions;
		return;
	}

	for (unsigned int j = 0; j < n; ++j)
	{
		if ( check_square(row, j) )
		{
			board[row][j] = 1;
			solve_queen_problem(row + 1);
			board[row][j] = 0;
		}
	}
}
